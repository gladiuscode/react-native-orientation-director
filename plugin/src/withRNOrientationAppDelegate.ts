import {
  type ConfigPlugin,
  type ExportedConfigWithProps,
  withAppDelegate,
} from '@expo/config-plugins';
import { type AppDelegateProjectFile } from '@expo/config-plugins/build/ios/Paths';
import { mergeContents } from '@expo/config-plugins/build/utils/generateCode';

export const withRNOrientationAppDelegate: ConfigPlugin = (config) => {
  return withAppDelegate(config, readAppDelegateFileAndUpdateContents);
};

async function readAppDelegateFileAndUpdateContents(
  config: ExportedConfigWithProps<AppDelegateProjectFile>
): Promise<ExportedConfigWithProps<AppDelegateProjectFile>> {
  const { modResults: appDelegateFile } = config;

  const worker = getCompatibleFileUpdater(appDelegateFile.language);
  appDelegateFile.contents = worker(appDelegateFile.contents);

  return config;
}

function getCompatibleFileUpdater(
  language: AppDelegateProjectFile['language']
): (originalContents: string) => string {
  switch (language) {
    case 'objc':
    case 'objcpp': {
      return objCFileUpdater;
    }
    case 'swift':
      return swiftFileUpdater;
    default:
      throw new Error(
        `Cannot add React Native Orientation Director code to AppDelegate of language "${language}"`
      );
  }
}

export function swiftFileUpdater(originalContents: string): string {
  const supportedInterfaceOrientationsForCodeBlock = `override func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
  return OrientationDirector.getSupportedInterfaceOrientationsForWindow()
}\n`;
  const rightBeforeLastClosingBrace = /}[^}]*$/g;

  const results = mergeContents({
    tag: '@react-native-orientation-director/supportedInterfaceOrientationsFor-implementation',
    src: originalContents,
    newSrc: supportedInterfaceOrientationsForCodeBlock,
    anchor: rightBeforeLastClosingBrace,
    offset: 0,
    comment: '// React Native Orientation Director',
  });

  return results.contents;
}

export function objCFileUpdater(originalContents: string): string {
  const libraryHeaderImportCodeBlock = '#import <OrientationDirector.h>\n';
  const rightBeforeAppDelegateImplementation = /@implementation\s+\w+/g;

  const headerImportMergeResults = mergeContents({
    tag: '@react-native-orientation-director/library-header-import',
    src: originalContents,
    newSrc: libraryHeaderImportCodeBlock,
    anchor: rightBeforeAppDelegateImplementation,
    offset: 0,
    comment: '// React Native Orientation Director',
  });

  const supportedInterfaceOrientationsForCodeBlock = `- (UIInterfaceOrientationMask)application:(UIApplication *)application supportedInterfaceOrientationsForWindow:(UIWindow *)window
{
  return [OrientationDirector getSupportedInterfaceOrientationsForWindow];
}\n`;
  const rightBeforeLastClosingEnd = /@end[^@]*$/g;

  const implementationMergeResults = mergeContents({
    tag: '@react-native-orientation-director/supportedInterfaceOrientationsFor-implementation',
    src: headerImportMergeResults.contents,
    newSrc: supportedInterfaceOrientationsForCodeBlock,
    anchor: rightBeforeLastClosingEnd,
    offset: 0,
    comment: '// React Native Orientation Director',
  });

  return implementationMergeResults.contents;
}
