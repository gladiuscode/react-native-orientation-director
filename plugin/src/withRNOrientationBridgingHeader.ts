import {
  type ConfigPlugin,
  type ExportedConfigWithProps,
  IOSConfig,
} from '@expo/config-plugins';
import { type AppDelegateProjectFile } from '@expo/config-plugins/build/ios/Paths';

import { withAppBridgingHeader } from './custom-mod/withBridgingHeader';

export const withRNOrientationBridgingHeader: ConfigPlugin = (config) => {
  return withAppBridgingHeader(config, readBridgingHeaderFileAndUpdateContents);
};

async function readBridgingHeaderFileAndUpdateContents(
  config: ExportedConfigWithProps<AppDelegateProjectFile>
) {
  const { projectRoot } = config.modRequest;
  if (isObjCTemplate(projectRoot)) {
    return config;
  }

  const { modResults: bridgingHeaderFile } = config;

  bridgingHeaderFile.contents = bridgingHeaderUpdater(
    bridgingHeaderFile.contents
  );

  return config;
}

export function bridgingHeaderUpdater(originalContents: string) {
  const libraryHeaderImport = '#import "OrientationDirector.h"';

  return originalContents.concat(`\n${libraryHeaderImport}`);
}

function isObjCTemplate(projectRoot: string) {
  const appDelegateFile = IOSConfig.Paths.getAppDelegate(projectRoot);
  return appDelegateFile.language !== 'swift';
}
