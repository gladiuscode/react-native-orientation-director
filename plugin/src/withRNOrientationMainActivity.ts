import {
  type ConfigPlugin,
  type ExportedConfigWithProps,
  withMainActivity,
} from '@expo/config-plugins';
import { type ApplicationProjectFile } from '@expo/config-plugins/build/android/Paths';
import { mergeContents } from '@expo/config-plugins/build/utils/generateCode';

export const withRNOrientationMainActivity: ConfigPlugin = (config) => {
  return withMainActivity(config, readMainActivityFileAndUpdateContents);
};

async function readMainActivityFileAndUpdateContents(
  config: ExportedConfigWithProps<ApplicationProjectFile>
): Promise<ExportedConfigWithProps<ApplicationProjectFile>> {
  const { modResults: mainActivityFile } = config;

  const worker = getCompatibleFileUpdater(mainActivityFile.language);
  mainActivityFile.contents = worker(mainActivityFile.contents);

  return config;
}

function getCompatibleFileUpdater(
  language: ApplicationProjectFile['language']
): (originalContents: string) => string {
  switch (language) {
    case 'kt':
      return ktFileUpdater;
    default:
      throw new Error(
        `Cannot add React Native Orientation Director code to MainActivity of language "${language}"`
      );
  }
}

export function ktFileUpdater(originalContents: string): string {
  const systemImportsContents =
    updateContentsWithSystemImports(originalContents);

  const libraryImportCodeBlock =
    'import com.orientationdirector.implementation.ConfigurationChangedBroadcastReceiver';
  const rightBeforeClassDeclaration = /class MainActivity/g;

  const importMergeResults = mergeContents({
    tag: '@react-native-orientation-director/library-import',
    src: systemImportsContents,
    newSrc: libraryImportCodeBlock,
    anchor: rightBeforeClassDeclaration,
    offset: 0,
    comment: '// React Native Orientation Director',
  });

  const onConfigurationChangedCodeBlock = `
  override fun onConfigurationChanged(newConfig: Configuration) {
    super.onConfigurationChanged(newConfig)

    val orientationDirectorCustomAction =
      packageName + "." + ConfigurationChangedBroadcastReceiver.CUSTOM_INTENT_ACTION

    val intent =
      Intent(orientationDirectorCustomAction).apply {
        putExtra("newConfig", newConfig)
        setPackage(packageName)
      }

    this.sendBroadcast(intent)
  }\n`;

  const rightBeforeLastClosingBrace = /super\.onCreate/g;
  const pasteInTheListJustAfterTheClosingBracket = 2;

  const implementationMergeResults = mergeContents({
    tag: '@react-native-orientation-director/supportedInterfaceOrientationsFor-implementation',
    src: importMergeResults.contents,
    newSrc: onConfigurationChangedCodeBlock,
    anchor: rightBeforeLastClosingBrace,
    offset: pasteInTheListJustAfterTheClosingBracket,
    comment: '// React Native Orientation Director',
  });

  return implementationMergeResults.contents;
}

function updateContentsWithSystemImports(originalContents: string) {
  const rightBeforeClassDeclaration = /class MainActivity/g;

  let possibleUpdatedContents = originalContents;
  possibleUpdatedContents = addIntentImportIfNecessary(possibleUpdatedContents);
  possibleUpdatedContents = addConfigurationImportIfNecessary(
    possibleUpdatedContents
  );

  return possibleUpdatedContents;

  function addIntentImportIfNecessary(_contents: string) {
    const systemIntentImportCodeBlock = 'import android.content.Intent';
    if (_contents.includes(systemIntentImportCodeBlock)) {
      return _contents;
    }

    const mergeResults = mergeContents({
      tag: '@react-native-orientation-director/system-intent-import',
      src: _contents,
      newSrc: systemIntentImportCodeBlock,
      anchor: rightBeforeClassDeclaration,
      offset: 0,
      comment: '// React Native Orientation Director',
    });

    return mergeResults.contents;
  }
  function addConfigurationImportIfNecessary(_contents: string) {
    const systemConfigurationImportCodeBlock =
      'import android.content.res.Configuration';
    if (possibleUpdatedContents.includes(systemConfigurationImportCodeBlock)) {
      return _contents;
    }

    const mergeResults = mergeContents({
      tag: '@react-native-orientation-director/system-configuration-import',
      src: possibleUpdatedContents,
      newSrc: systemConfigurationImportCodeBlock,
      anchor: rightBeforeClassDeclaration,
      offset: 0,
      comment: '// React Native Orientation Director',
    });

    return mergeResults.contents;
  }
}
