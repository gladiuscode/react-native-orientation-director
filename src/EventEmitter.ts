import { Platform, type EmitterSubscription } from 'react-native';
import Module, { ModuleEventEmitter } from './module';
import Event from './types/Event.enum';
import type { OrientationEvent } from './types/OrientationEvent.interface';
import type { LockedEvent } from './types/LockedEvent.interface';

class EventEmitter {
  static addDeviceOrientationDidChangeListener(
    callback: (orientation: OrientationEvent) => void
  ) {
    let listener = ModuleEventEmitter.addListener(
      Event.DeviceOrientationDidChange,
      callback
    );

    if (Platform.OS !== 'android') {
      return listener;
    }

    const listenerCount = ModuleEventEmitter.listenerCount(
      Event.DeviceOrientationDidChange
    );

    if (listenerCount === 0) {
      Module.enableOrientationSensors();
    }

    return EventEmitter.createDeviceOrientationListenerProxy(listener);
  }

  static addInterfaceOrientationDidChangeListener(
    callback: (orientation: OrientationEvent) => void
  ) {
    return ModuleEventEmitter.addListener(
      Event.InterfaceOrientationDidChange,
      callback
    );
  }

  static addLockDidChangeListener(callback: (event: LockedEvent) => void) {
    return ModuleEventEmitter.addListener(Event.LockDidChange, callback);
  }

  private static createDeviceOrientationListenerProxy(
    listener: EmitterSubscription
  ) {
    const handler: ProxyHandler<EmitterSubscription> = {
      get(target, propertyKey, receiver) {
        if (propertyKey === 'remove') {
          disableOrientationSensorsIfLastListener();
        }
        return Reflect.get(target, propertyKey, receiver);
      },
    };

    return new Proxy(listener, handler);

    function disableOrientationSensorsIfLastListener() {
      const listenerCount = ModuleEventEmitter.listenerCount(
        Event.DeviceOrientationDidChange
      );

      if (listenerCount === 1) {
        Module.disableOrientationSensors();
      }
    }
  }
}

export default EventEmitter;
