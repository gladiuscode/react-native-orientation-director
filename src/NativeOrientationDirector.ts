import type { TurboModule } from 'react-native';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  getInterfaceOrientation(): Promise<number>;
  getDeviceOrientation(): Promise<number>;
  lockTo(orientation: number): void;
  unlock(): void;
  isLocked(): boolean;
  isAutoRotationEnabled(): boolean;
  resetSupportedInterfaceOrientations(): void;

  addListener: (eventType: string) => void;
  removeListeners: (count: number) => void;
}

export default TurboModuleRegistry.getEnforcing<Spec>('OrientationDirector');
