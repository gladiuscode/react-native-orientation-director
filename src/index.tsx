export { Orientation } from './types/Orientation.enum';

import useDeviceOrientation from './hooks/useDeviceOrientation.hook';
export { useDeviceOrientation };

import useInterfaceOrientation from './hooks/useInterfaceOrientation.hook';
export { useInterfaceOrientation };

import useIsInterfaceOrientationLocked from './hooks/useIsInterfaceOrientationLocked.hook';
export { useIsInterfaceOrientationLocked };

import RNOrientationDirector from './RNOrientationDirector';
export default RNOrientationDirector;
