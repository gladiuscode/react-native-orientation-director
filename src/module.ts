import { NativeEventEmitter, NativeModules, Platform } from 'react-native';
import type { Spec } from './NativeOrientationDirector';

const LINKING_ERROR =
  `The package 'react-native-orientation-director' doesn't seem to be linked. Make sure: \n\n` +
  Platform.select({ ios: "- You have run 'pod install'\n", default: '' }) +
  '- You rebuilt the app after installing the package\n' +
  '- You are not using Expo Go\n';

// @ts-expect-error
const isTurboModuleEnabled = global.__turboModuleProxy != null;

const Module = isTurboModuleEnabled
  ? require('./NativeOrientationDirector').default
  : NativeModules.OrientationDirector;

const OrientationDirectorModule = Module
  ? Module
  : new Proxy(
      {},
      {
        get() {
          throw new Error(LINKING_ERROR);
        },
      }
    );

export const ModuleEventEmitter = new NativeEventEmitter(Module);

export default OrientationDirectorModule as Spec;
