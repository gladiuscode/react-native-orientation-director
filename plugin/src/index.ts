import { type ConfigPlugin, withPlugins } from '@expo/config-plugins';

import { withAppBridgingHeaderMod } from './custom-mod/withBridgingHeader';
import { withRNOrientationAppDelegate } from './withRNOrientationAppDelegate';
import { withRNOrientationBridgingHeader } from './withRNOrientationBridgingHeader';

/**
 * So, expo config plugin are awesome and the documentation is well written, but I still needed to look around to see
 * how other projects actually modify the AppDelegate. I've found react-native-firebase to implement a plugin config
 * that changes the AppDelegate, so I'll leave their link as reference:
 * https://github.com/invertase/react-native-firebase/blob/main/packages/app/plugin/src/ios/appDelegate.ts
 *
 * Kudos to them, because this stuff is hard!
 *
 * @param config
 * @param name
 */
const withRNOrientationDirector: ConfigPlugin<{ name?: string }> = (
  config,
  { name = 'react-native-orientation-director' } = {}
) => {
  config.name = name;
  return withPlugins(config, [
    withRNOrientationAppDelegate,
    withRNOrientationBridgingHeader,
    withAppBridgingHeaderMod,
  ]);
};

export default withRNOrientationDirector;
