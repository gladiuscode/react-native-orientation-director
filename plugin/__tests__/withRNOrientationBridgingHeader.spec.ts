import * as fs from 'node:fs';
import * as path from 'node:path';

import { bridgingHeaderUpdater } from '../src/withRNOrientationBridgingHeader';

describe('withRNOrientationBridgingHeader', function () {
  beforeEach(function () {
    jest.resetAllMocks();
  });

  it('updates the App Bridging-Header with the header import', async function () {
    const bridgingHeaderPath = path.join(
      __dirname,
      './fixtures/Bridging-Header.h'
    );
    const bridgingHeader = await fs.promises.readFile(
      bridgingHeaderPath,
      'utf-8'
    );

    const result = bridgingHeaderUpdater(bridgingHeader);
    expect(result).toMatchSnapshot();
  });
});
