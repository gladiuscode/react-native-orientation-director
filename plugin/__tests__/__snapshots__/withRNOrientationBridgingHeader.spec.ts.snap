// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRNOrientationBridgingHeader updates the App Bridging-Header with the header import 1`] = `
"//
//  Use this file to import your target's public headers that you would like to expose to Swift.
//

#import "OrientationDirector.h""
`;
